### ISSUE TITLE
createFragmentContainer massively slowed down by irrelevant data

### ISSUE THREAD

**Issue by piotrblasiak** [facebook/relay#2431]
I am fetching massive amounts of data using relay with a graphql query like:

`query SomeQuery {
    myType {
        id
        name
        entries {
            ...SomeFragment_entry
        }
    }
}`

The bulk of the object will be in `entries` which can be 10s of thousands of elements.

And then I have a fragment container that only needs the id & name data:

`createFragmentContainer(Header, graphql`
    fragment HeaderFragment_myType on MyType {
        id
        name
    }
`)`

Now whenever I re-render that Header component, it is super slow. I suspected this could have something to do with the createFragmentContainer and indeed - if I remove it the problem disappears. It seems as if the masking process for some reason slows down with the size of the `entries` field in the myType object it is masking, even tho it is not at all in the fragment.
I verified (and temporarily fixed) this problem by omiting the entries prop from the myType object given to the fragment:

`<Header myType={omit(myType, "entries")}/>`

And after that it was working wonderfully.

I suspect not that many people are using relay for massive data sets like me, and that this might not be a noticeable bug but for me it is a problem and I suspect this might also make relay a lot faster for a lot of people if fixed.

**Comment by kikoseijo** [comment]
Indeed its rare,

What about your component? stateless?
How are you rendering of this component?

**Comment by piotrblasiak** [comment]
It does not seem to matter what the component is doing - it is the createFragmentContainer that adds a lot of overhead - and I realised that the only upside of using it was the masking & data checking which I´ll have to live without for now. So instead I am now using fragments but with @relay(mask: false) so that it works without createFragmentContainer.

**Comment by kikoseijo** [comment]
Can you just pass the props to the component directly? this will avoid you to use the  @relay(mask: false) 

Im my opinion, when a simple component does not required a lot of data and its always going to be used inside same controller, `Props` are faster and easier.

Im not sure if a createFragmentContainer with the Relay thing takes overhead or adds.

**Comment by piotrblasiak** [comment]
When using fragments I need to use @relay(mask: false) so that I can pass the data as props directly without createFragmentContainer, no?

**Comment by kikoseijo** [comment]
No you don't! use as it in normal, once your prop its at your component pass it down like normal. Using he @mask(false) of course will save you passing down the props to the component manually.

Not sure if by making things visible makes too much overhead, but as you doing large data sets, should go this way instead of masking off.

**Comment by piotrblasiak** [comment]
Not sure if we´re talking about the same thing @kikoseijo cause when I´m not unmasking it does not work - and if I´m debugging I can clearly see that there is a __fragments prop on the data object and not the fields from the fragment.

**Comment by kikoseijo** [comment]
Probably not @piotrblasiak 
Cant help you with the Relay, never had that problem, what Im trying to say its that in my opinion its easier to pass down a property on a child component using the declaration method `<ChildComponent myprop={this.props.whatever} />` than using a `createFragmentContainer`

By doing so you don't need to disable masking.

**Comment by mike-marcacci** [comment]
@kikoseijo - I think that @piotrblasiak is pointing out that fields requested in a child fragment are inaccessible without `@relay(mask: false)`, which is correct.

I tried to ditch `createFragmentContainer` due to these performance issues and the myriad bugs that result from the use of react context to pass relay variables to fragment containers... but I am held up by #2118.

**Comment by piotrblasiak** [comment]
I´m glad(?) to hear that more people than just me are having performance issues with fragment containers. Is this something raised somewhere else and that the relay team is aware of?

**Comment by sibelius** [comment]
can anyone provide a benchmark on this?

**Comment by jstejada** [comment]
> the myriad bugs that result from the use of react context to pass relay variables to fragment containers

FWIW, we're no longer going to pass variables through context

**Comment by jstejada** [comment]
A benchmark or a Chrome Dev Tools Profile would be really helpful here.

**Comment by sorenhoyer** [comment]
Experienced this issue as I was refactoring a "funky" component that rendered a recharts line chart with 5000 datapoints. The problem was it had a mouseover eventhandler where og would dispatch (useReducer) every time, meaning a new render each time. Not the best code. However the dev did get away with it, as it was only when I refactored it to use 2 fragments (useFragment) that it "crashed" (performance dropped through the roof). To be fair we're still on relay 7 that shipped with hooks (accidentally).

**Comment by sibelius** [comment]
Migrate to relay hooks, if the issue still valid, open another issue
